arbor/morph/locset.cpp

```diff
@@ -137,6 +143,82 @@ std::ostream& operator<<(std::ostream& o, const named_& x) {
     return o << "(locset \"" << x.name << "\")";
 }
 
+// Uniform locset.
+
+struct uniform_ {
+    region reg;
+    unsigned left;
+    unsigned right;
+    uint64_t seed;
+};
+
+locset uniform(arb::region reg, unsigned left, unsigned right, uint64_t seed) {
+    return locset(uniform_{reg, left, right, seed});
+}
+
+mlocation_list thingify_(const uniform_& u, const mprovider& p) {
+    mlocation_list L;
+
+    typedef r123::Threefry2x64 cbrng;
+    auto morpho = p.morphology();
+    auto embed = p.embedding();
+
+    // Thingify the region and store relevant data
+    auto reg_cables = thingify(u.reg, p);
+
+    std::vector<double> lengths_bounds;
+    auto lengths_part = util::make_partition(lengths_bounds,
+                                       util::transform_view(reg_cables, [&embed](const auto& c) {
+                                           return embed.integrate_length(c);
+                                       }));
+
+    auto region_length = lengths_part.bounds().second;
+
+    // Generate uniform random positions along the extent of the full region
+    cbrng::key_type key = {{u.seed}};
+    cbrng::ctr_type ctr = {{0,0}};
+    cbrng g;
+
+    std::vector<double> random_pos;
+    auto left  = u.left%2  ? u.left  - 1 : u.left;
+    auto right = u.right%2 ? u.right + 1 : u.right;
+
+    for (unsigned i = left; i < right; i+=2) {
+        ctr[0] = i/2;
+        cbrng::ctr_type rand = g(ctr, key);
+        random_pos.push_back(r123::u01<double>(rand[0])*region_length);
+        random_pos.push_back(r123::u01<double>(rand[1])*region_length);
+    }
+
+    if (u.left%2) {
+        random_pos.erase(random_pos.begin());
+    }
+    if (u.right%2) {
+        random_pos.pop_back();
+    }
+    util::sort(random_pos);
+
+    // Match random_extents to cables and find position on the associated branch
+    unsigned cable_idx = 0;
+    auto range = lengths_part[cable_idx];
+
+    for (auto e: random_pos) {
+        while (e > range.second) {
+            range = lengths_part[++cable_idx];
+        }
+        auto cable = reg_cables[cable_idx];
+        auto pos_on_cable = (e - range.first)/(range.second - range.first);
+        auto pos_on_branch = (cable.dist_pos - cable.prox_pos)*pos_on_cable + cable.prox_pos;
+        L.push_back({cable.branch, pos_on_branch});
+    }
+
+    return L;
+}
+
+std::ostream& operator<<(std::ostream& o, const uniform_& u) {
+    return o << "(uniform from region: \"" << u.reg << "\"; using seed: " << u.seed
+             << "; range: {" << u.left << ", " << u.right << "})";
+}
```

Comment:
Just in terms of the user interface, I am wondering if it makes sense to include the index `right` in the range of generated numbers, i.e., have `uniform` specify a closed interval, like `std::uniform_distribution<int>` for example, rather than a half-open one.